File: problems/Python/16. 3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        res = nums[0] + nums[1] + nums[2]

        for i in range(len(nums) - 2):
            l, r = i + 1, len(nums) - 1
            while l < r:
                csum = nums[i] + nums[l] + nums[r]
                
                if abs(csum - target) < abs(res - target):
                    res = csum
                
                if csum < target:
                    l += 1
                elif csum > target:
                    r -= 1
                else:
                    return csum
                    
        return res

File: problems/C++/16. 3Sum Closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int res = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < nums.size() - 2; ++i) {
            int l = i + 1;
            int r = nums.size() - 1;
            while (l < r) {
                int csum = nums[i] + nums[l] + nums[r];
                if (abs(csum - target) < abs(res - target)) {
                    res = csum;
                }

                if (csum < target) {
                    l++;
                } else if (csum > target) {
                    r--;
                } else {
                    return csum;
                }
            }
        }
        return res;
    }
};